### PR TITLE
Implementing safety rules for the "ref readonly" parameters and returns

### PR DESCRIPTION
**Pull request by VSadov**
- expression capabilities checks are refactored into a separate file.  
We had these checks scattered around the binder and it was not easy to follow.

- makes expression capabilities enum `BindValueKind` a `[Flags]` enum.
Turns out all the capabilities could be expressed in terms of just 4 values -

 1. Expression is readable   
 2. Expression is writeable  
 3. Expression is a reference to a location   
     (can be bound to a standalone variable in a referentially-transparent way.)
 4. Expression is returnable
     (free of references to the local stack frame)

We still need all the other values of `BindValueKind`, for the quality of error reporting.  
However most of the validation becomes much simpler if it is done in terms of the 4 base kinds.

- implements actual safety checks for "ref readonly" parameters and returns as in [the proposal](https://github.com/dotnet/csharplang/blob/master/proposals/readonly-ref.md).

The most Interesting things to note

1. "in" parameters are                    { Readable | Returnable | Refer to a location }
2. "ref readonly" returns are          { Readable | Returnable if all args are | Refer to a location}
3. readonly fields outside of the constructor { Readable | Returnable (*) | Refer to a location }   
4. (*) returnability of struct instance fields defers to the receiver.
5. properties are particularly tricky - capabilities depend on whether there is an accessible setter, getter, and whether it returns by value, by reference or by readonly reference.
6. returnability of "ref readonly" indexers depends on returnability of the index expressions.

UNDONE:
- ternary refs should work correctly, but need more tests, since they have been just merged in. (tracked as  #17797)
- returnability of indexers/methods with "in" parameters - the checks are implemented, but not tested since we cannot yet call such methods/indexers. (tracked as #17798)

### CODE REVIEW

**Comment by VSadov** [comment]
@OmarTawfik  @dotnet/roslyn-compiler - please review.

**Comment by VSadov** [comment]
It used to pass all the tests. The merge conflict is with just merged "ternary ref". Should be trivial to resolve.

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - ping

**Comment by VSadov** [comment]
test windows_release_vs-integration_prtest please

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 1]
```diff
@@ -1462,6 +1462,13 @@ internal enum ErrorCode
 
         ERR_Merge_conflict_marker_encountered = 8300,
 
+        ERR_RefReadonlyNotField = 8206,
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: >ERR_RefReadonlyNotField [](start = 8, length = 23)

Please add after 8404

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -5815,9 +5751,14 @@ private static void CombineExtensionMethodArguments(BoundExpression receiver, An
                     if (IsNonMoveableVariable(receiver, out accessedLocalOrParameterOpt) == isFixedStatementExpression && !isInsideNameof)
                     {
                         Error(diagnostics, isFixedStatementExpression ? ErrorCode.ERR_FixedNotNeeded : ErrorCode.ERR_FixedBufferNotFixed, node);
-                        hasErrors = true;
+                        hasErrors = hasError = true;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Do we need to assign to `hasErrors` here? We don't assign to anywhere else.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -4796,7 +4814,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Cannot return a parameter by reference '{0}' because it is not a ref or out parameter</value>
   </data>
   <data name="ERR_RefReturnParameter2" xml:space="preserve">
-    <value>Cannot return or a member of parameter '{0}' by reference because it is not a ref or out parameter</value>
+    <value>Cannot return by reference a member of parameter '{0}' by reference because it is not a ref or out parameter</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: `by reference` is included twice. Should this be `"Cannot return a member of parameter '{0}' by reference ..."`?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 4]
```diff
@@ -4796,7 +4814,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Cannot return a parameter by reference '{0}' because it is not a ref or out parameter</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Perhaps `"Cannot return parameter '{0}' by reference ..."`

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 5]
```diff
@@ -98,5 +98,183 @@ class Program
...
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (6,9): error CS8408: Cannot assign to variable 'ref readonly int' because it is a readonly variable
+                //         arg1 = 1;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "arg1").WithArguments("variable", "ref readonly int").WithLocation(6, 9),
+                // (7,9): error CS8409: Cannot assign to a member of variable 'ref readonly (int Alice, int Bob)' because it is a readonly variable
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: `Cannot assign to a member of variable 'arg2' ...`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 6]
```diff
@@ -98,5 +98,183 @@ class Program
...
+                //         arg2.Alice = 2;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField2, "arg2.Alice").WithArguments("variable", "ref readonly (int Alice, int Bob)").WithLocation(7, 9),
+                // (9,9): error CS8408: Cannot assign to variable 'ref readonly int' because it is a readonly variable
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: `Cannot assign to variable 'arg1' ...`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 7]
```diff
@@ -98,5 +98,183 @@ class Program
...
+    {
+        ref var y = ref arg1;
+        ref int a = ref arg2.Alice;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (6,25): error CS8406: Cannot use variable 'ref readonly int' as a ref or out value because it is a readonly variable
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: `Cannot use variable 'arg1' as a ref or out ...`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 8]
```diff
@@ -98,5 +98,183 @@ class Program
...
+                //         ref var y = ref arg1;
+                Diagnostic(ErrorCode.ERR_RefReadonlyNotField, "arg1").WithArguments("variable", "ref readonly int").WithLocation(6, 25),
+                // (7,25): error CS8407: Members of variable 'ref readonly (int Alice, int Bob)' cannot be used as a ref or out value because it is a readonly variable
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: `Members of variable 'arg2' cannot be used as a ref or out ...`?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 9]
```diff
@@ -0,0 +1,1135 @@
...
+                    {
+                        Error(diagnostics, ErrorCode.ERR_UnboxNotLValue, node);
+                        return false;
+                    }
+                    break;
+
+                case BoundKind.ArrayAccess:
+                case BoundKind.PointerIndirectionOperator:
+                case BoundKind.PointerElementAccess:
+                    // array elements and pointer dereferencing are readwrite varaibles
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Typo: `variables`

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 10]
```diff
@@ -0,0 +1,1135 @@
...
+
+                    // byref conditional defers to its operands
+                    if (conditional.IsByRef && 
+                        (CheckValueKind(conditional.Consequence.Syntax, conditional.Consequence, valueKind, checkingReceiver: false, diagnostics: diagnostics) &
+                        CheckValueKind(conditional.Alternative.Syntax, conditional.Alternative, valueKind, checkingReceiver: false, diagnostics: diagnostics)))
+                    {
+                        return true;
+                    }
+
+                    // reprot standard lvalue error
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Typo: `report`

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 11]
```diff
@@ -1116,623 +1116,6 @@ private BoundExpression GetFixedLocalCollectionInitializer(BoundExpression initi
...
-                return false;
-            }
-
-            if (kind == BindValueKind.AddressOf)
-            {
-                Error(diagnostics, ErrorCode.ERR_AddrOnReadOnlyLocal, tree);
-                return false;
-            }
-
-            ErrorCode[] ReadOnlyLocalErrors =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Should this be a `static readonly` field?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 12]
```diff
@@ -0,0 +1,1135 @@
...
+                return;
+            }
+
+            if (kind == BindValueKind.AddressOf)
+            {
+                Error(diagnostics, ErrorCode.ERR_AddrOnReadOnlyLocal, node);
+                return;
+            }
+
+            ErrorCode[] ReadOnlyLocalErrors =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Should this be a `static readonly` field?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 13]
```diff
@@ -0,0 +1,1135 @@
...
+
+            // It's clearer to say that the address can't be taken than to say that the field can't be modified
+            // (even though the latter message gives more explanation of why).
+            if (kind == BindValueKind.AddressOf)
+            {
+                Error(diagnostics, ErrorCode.ERR_InvalidAddrOp, node);
+                return;
+            }
+
+            ErrorCode[] ReadOnlyErrors =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: `static readonly` field?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 14]
```diff
@@ -0,0 +1,1135 @@
...
+            // (even though the latter message gives more explanation of why).
+            if (kind == BindValueKind.AddressOf)
+            {
+                Error(diagnostics, ErrorCode.ERR_InvalidAddrOp, node);
+                return;
+            }
+
+            var symbolKind = symbol.Kind.Localize();
+
+            ErrorCode[] ReadOnlyErrors =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: `static readonly` field?

**Pull request merged**